common/cpp/src/flutter_screen_capture.cc

```diff
@@ -8,86 +8,213 @@
-  SourceList sources;
+ FlutterScreenCapture::FlutterScreenCapture(FlutterWebRTCBase *base): base_(base) {
+std::string event_channel =
+      "FlutterWebRTC/desktopSourcesEvent";
+  event_channel_.reset(new EventChannel<EncodableValue>(
+    base_->messenger_, event_channel, &StandardMethodCodec::GetInstance()));
+
+  auto handler = std::make_unique<StreamHandlerFunctions<EncodableValue>>(
+		  [&](
+			  const flutter::EncodableValue* arguments,
+			  std::unique_ptr<flutter::EventSink<flutter::EncodableValue>>&& events)
+		  -> std::unique_ptr<StreamHandlerError<flutter::EncodableValue>> {
+	  event_sink_ = std::move(events);
+	  return nullptr;
+  },
+		  [&](const flutter::EncodableValue* arguments)
+	  -> std::unique_ptr<StreamHandlerError<flutter::EncodableValue>> {
+	  event_sink_ = nullptr;
+	  return nullptr;
+  });
+
+  event_channel_->SetStreamHandler(std::move(handler));
+ }
 
-  sources = base_->desktop_device_->EnumerateScreens();
-  
-  EncodableList list;
-  for (const auto& source : sources.std_vector()) {
-    // std::cout << " id: " << source.id.std_string() << " title: " << source.title.std_string() << " type: " << source.type << std::endl;
-    EncodableMap info;
-    info[EncodableValue("id")] = EncodableValue(source.id.std_string());
-    info[EncodableValue("name")] = EncodableValue(source.title.std_string());
-    info[EncodableValue("type")] = EncodableValue(source.type);
-    list.push_back(EncodableValue(info));
-  }
-
-  result->Success(EncodableValue(list));
-}
-
-void FlutterScreenCapture::EnumerateWindows(std::unique_ptr<MethodResult<EncodableValue>> result) {
-  SourceList sources;
-
-  sources = base_->desktop_device_->EnumerateWindows();
-  
-  EncodableList list;
-  for (const auto& source : sources.std_vector()) {
-    // std::cout << " id: " << source.id.std_string() << " title: " << source.title.std_string() << " type: " << source.type << std::endl;
-    EncodableMap info;
-    info[EncodableValue("id")] = EncodableValue(source.id.std_string());
-    info[EncodableValue("name")] = EncodableValue(source.title.std_string());
-    info[EncodableValue("type")] = EncodableValue(source.type);
-    list.push_back(EncodableValue(info));
+void FlutterScreenCapture::GetDesktopSources(const EncodableList &types, std::unique_ptr<MethodResult<EncodableValue>> result) {
+  EncodableList sources;
+  sources_.clear();
+  size_t size = types.size();
+  for (size_t i = 0; i < size; i++) {
+    std::string type = GetValue<std::string>(types[i]);
+    if (type == "window") {
+      scoped_refptr<RTCDesktopMediaList> window;
+      auto it = medialist_.find(kWindow);
+      if (it != medialist_.end()) {
+        window = (*it).second;
+      } else {
+        window = base_->desktop_device_->GetDesktopMediaList(kWindow);
+        window->RegisterMediaListObserver(this);
+        medialist_[kWindow] = window;
+      }
+      window->UpdateSourceList(true);
```

Comment:
We pass in `true` here, but everywhere we pass in false? Is there something specific to this particular code path?

Reply:
Here is the method in libwebrtc. When true, it will rebuild the entire source list. When it is false, it will use add or remove to refresh the list. So I will give it false when updateDesktopSources
```cpp
virtual int32_t UpdateSourceList(bool force_reload = false, bool get_thumbnail = true) = 0;
```

Reply:
So if it's false, the `OnMediaSourceAdded/Removed` callbacks will get called, while if it's true, those callbacks won't be called?

Reply:
If it is true, all items will be removed and OnMediaSourceRemoved will be called, when false, only the window sources that are invalidated or closed will be removed